research/minigo/minigo.py

```diff
@@ -305,38 +278,189 @@ def evaluate(trained_models_dir, black_model_name, white_model_name,
   params = model_params.MiniGoParams()
-  k = utils.round_power_of_two(board_size ** 2 / 3)
+  k = utils.round_power_of_two(FLAGS.board_size ** 2 / 3)
   params.num_filters = k  # Number of filters in the convolution layer
   params.fc_width = 2 * k  # Width of each fully connected layer
-  params.num_shared_layers = board_size  # Number of shared trunk layers
-  params.board_size = board_size  # Board size
+  params.num_shared_layers = FLAGS.board_size  # Number of shared trunk layers
+  params.board_size = FLAGS.board_size  # Board size
 
   # How many positions can fit on a graphics card. 256 for 9s, 16 or 32 for 19s.
-  if FLAGS.board_size == 9:
-    params.batch_size = 256
+  if FLAGS.batch_size is None:
+    if FLAGS.board_size == 9:
+      params.batch_size = 256
+    else:
+      params.batch_size = 32
   else:
-    params.batch_size = 32
+    params.batch_size = FLAGS.batch_size
+
   return params
 
 
+def _prepare_selfplay(
+    model_name, trained_models_dir, selfplay_dir, holdout_dir, sgf_dir, params):
+  """Set directories and load the network for selfplay.
+
+  Args:
+    model_name: The name of the model for self-play
+    trained_models_dir: Directories where the completed generations/models are.
+    selfplay_dir: Where to write the games. Set as 'base_dir/data/selfplay/'.
+    holdout_dir: Where to write the holdout data. Set as
+      'base_dir/data/holdout/'.
+    sgf_dir: Where to write the sgf (Smart Game Format) files. Set as
+      'base_dir/sgf/'.
+    params: An object of hyperparameters for the model.
+
+  Returns:
+    The directories and network model for selfplay.
+  """
+  print('Playing a game with model {}'.format(model_name))
+  # Set paths for the model with 'model_name'
+  model_path = os.path.join(trained_models_dir, model_name)
+  output_dir = os.path.join(selfplay_dir, model_name)
+  holdout_dir = os.path.join(holdout_dir, model_name)
+  # clean_sgf is to write sgf file without comments.
+  # full_sgf is to write sgf file with comments.
+  clean_sgf = os.path.join(sgf_dir, model_name, 'clean')
+  full_sgf = os.path.join(sgf_dir, model_name, 'full')
+
+  _ensure_dir_exists(output_dir)
+  _ensure_dir_exists(holdout_dir)
+  _ensure_dir_exists(clean_sgf)
+  _ensure_dir_exists(full_sgf)
+  selfplay_dirs = {
+      'output_dir': output_dir,
+      'holdout_dir': holdout_dir,
+      'clean_sgf': clean_sgf,
+      'full_sgf': full_sgf
+  }
+  # cache the network model for self-play
+  with utils.logged_timer('Loading weights from {} ... '.format(model_path)):
+    network = dualnet.DualNetRunner(model_path, params)
+  return selfplay_dirs, network
+
+
+def run_single_step_only(params, dirs):
+  """Run each individual step only.
+
+  Args:
+    params: An object of hyperparameters for the model.
+    dirs: An object of directories used in each step.
+
+  Raises:
+    ValueError: when neither of the black model and the white model is provided
+    for evaluation.
+  """
+  # bootstrap only
+  if FLAGS.bootstrap_only:
```

Comment:
Why give the option to run one only, and not just the ability to run multiple selected parts? Ie, have skip_bootstrap or skip_selfplay etc. so that if I want to bootstrap and selfplay, I can. That is no extra work or code, but gives the user more flexibility in running modes.

Reply:
Looking further down below, we are introducing a ton of flags with interactions and dependencies that are not clear, and this seems like a bad idea. Not to mention that there is now essentially a duplicated main function. Why aren't there just booleans like flags.evaluation and flags.validation below? We should simply skip a step if it is not requested, not change the path of execution. As to the extra flags like model_names, do we actually need that? Or is the directory structure sufficient? If we don't have specific use-cases, we should leave out, and if we do, we should consider requiring a config file, as this is a ton of flags.

Reply:
Thanks for the these comments, Karmel! This is exactly something that needs your inputs. 
1. I think skipping some specific step is okay. But according to the discussions with Victor last week, my understanding is that he cares more about parallelization of selfplay and train steps with multiple GPU, as those two steps take long. As for other steps, like bootstrap, we need it only once; gather is executed once after selfplay; validation is supposed to be called once to validate the model-just-trained; Not sure about evaluation, it may be called multiple times for model comparison. I think how to set flags mainly depends on how the users use it. @bitfort As you are the main user now, could you give some suggestions on it?
2. For model_names with selfplay, if users always want the latest model to generate selfplay data, the directory structure is totally enough; Otherwise, user has to specify a model for selfplay. Like you said, how we design these flags is up to the use-cases. When working on this part, I also feel like we have too many flags which may make users confused. Any advice from user's perspective @bitfort ?